src/main/scala/mesosphere/marathon/api/akkahttp/v2/EventsController.scala

```diff
@@ -70,76 +61,44 @@ class EventsController(
         }
       }
     }
-  }
 }
 
 object EventsController extends StrictLogging {
-
   /**
-    * Represents a graph stage that implements a Source[MarathonEvent] which:
-    * - registers to the given EventStream and pushes all elements downstream.
-    * - buffers maxMessages elements. The stream fails, if the buffer is full.
-    * - is leader aware. The stream completes, if this instance abdicates.
-    * - publishes an EventStreamAttached, if the stream is materialized
-    * - publishes an EventStreamDetached, if the stream is completed or failed.
+    * An event source which:
+    * - Yields all MarathonEvent's for the event bus whilst a leader.
+    * - is leader aware. The stream completes if this instance abdicates.
+    * - publishes an EventStreamAttached when the stream is materialized
+    * - publishes an EventStreamDetached when the stream is completed or fails
     * @param eventStream the event stream to subscribe to
     * @param bufferSize the size of events to buffer, if there is no demand.
     * @param remoteAddress the remote address
     */
-  class EventStreamSourceGraph(
-      eventStream: EventStream,
-      bufferSize: Int,
-      remoteAddress: RemoteAddress,
-      leaderStateEvents: Source[LeadershipState, Cancellable]) extends GraphStage[SourceShape[MarathonEvent]] with StrictLogging {
-    private[this] val out: Outlet[MarathonEvent] = Outlet("EventsController.EventPublisher.out")
-    override val shape: SourceShape[MarathonEvent] = SourceShape.of(out)
+  def eventStreamLogic(eventStream: EventStream, leaderEvents: Source[LeadershipState, Any], bufferSize: Int, remoteAddress: RemoteAddress) = {
 
-    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
-      val messages: mutable.Queue[MarathonEvent] = mutable.Queue.empty
-      var actorRef: Option[StageActor] = None
-      var electionSubscription: Cancellable = _
+    // Used to propagate a "stream close" signal when we see a LeadershipState.Standy event
+    val leaderLossKillSwitch =
+      leaderEvents.collect { case evt: LeadershipState.Standby => evt }
 
-      setHandler(out, new OutHandler {
-        override def onPull(): Unit = if (messages.nonEmpty) push(out, messages.dequeue())
-      })
-
-      override def preStart(): Unit = {
-        super.preStart()
-        val actor = getStageActor(receive)
-        val doCompleteStage = getAsyncCallback[Unit] { _ => completeStage() }
-        electionSubscription = leaderStateEvents.to(Sink.foreach({
-          case _: LeadershipState.Standby => doCompleteStage.invoke(())
-          case _ => ()
-        })).run()(this.materializer)
+    EnrichedSource.eventBusSource(classOf[MarathonEvent], eventStream, bufferSize, OverflowStrategy.fail)
+      .via(EnrichedFlow.stopOnFirst(leaderLossKillSwitch))
+      .watchTermination() { (m, completed) =>
         eventStream.publish(EventStreamAttached(remoteAddress = remoteAddress.toString()))
```

Comment:
Do you think it worth moving this to `.mapMaterializedValue` part?

Reply:
After thinking about this, I think you're right. It is a bit confusing to hijack the materialization reducer function. Better to do `.watchTermination()(Keep.both).mapMaterializedValue { m => ... }`